blockstack-vm/src/database.rs

```diff
@@ -0,0 +1,122 @@
+use std::collections::HashMap;
+
+use errors::Error;
+use InterpreterResult;
+use types::{Value, TypeSignature, TupleTypeSignature, AtomTypeIdentifier};
+
+pub trait DataMap {
+    fn fetch_entry(&self, key: &Value) -> InterpreterResult;
+    fn set_entry(&mut self, key: Value, value: Value) -> Result<(), Error>;
+    fn insert_entry(&mut self, key: Value, value: Value) -> InterpreterResult;
+    fn delete_entry(&mut self, key: &Value) -> InterpreterResult;
+}
+
+pub trait ContractDatabase {
+    fn get_data_map(&mut self, map_name: &str) -> Option<&mut DataMap>;
+    fn create_map(&mut self, map_name: &str, key_type: TupleTypeSignature, value_type: TupleTypeSignature);
+}
+
+pub struct MemoryDataMap {
+    map: HashMap<Value, Value>,
+    key_type: TypeSignature,
+    value_type: TypeSignature
+}
+
+pub struct MemoryContractDatabase {
+    maps: HashMap<String, MemoryDataMap>,
+}
+
+impl MemoryDataMap {
+    pub fn new(key_type: TupleTypeSignature,
+               value_type: TupleTypeSignature) -> MemoryDataMap {
+        MemoryDataMap {
+            map: HashMap::new(),
+            key_type: TypeSignature::new(
+                AtomTypeIdentifier::TupleType(key_type), 0),
+            value_type: TypeSignature::new(
+                AtomTypeIdentifier::TupleType(value_type), 0)
+        }
+    }
+}
+
+impl MemoryContractDatabase {
+    pub fn new() -> MemoryContractDatabase {
+        MemoryContractDatabase { maps: HashMap::new() }
+    }
+}
+
+impl ContractDatabase for MemoryContractDatabase {
+    fn get_data_map(&mut self, map_name: &str) -> Option<&mut DataMap> {
```

Comment:
This feels non-ergonomic to me.  Aren't there cases where you'd want an immutable reference to the map?  i.e. just to read it?

Reply:
Yep, I'll rename this function to `get_mut_data_map` and add `get_data_map` for an immutable reference.